File: QuizApp/src/QuizApp.BLL/Services/UrlService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using AutoMapper;

using QuizApp.BLL.Dto.TestResult;
using QuizApp.BLL.Dto.Url;
using QuizApp.BLL.Interfaces;
using QuizApp.Data.Interfaces;
using QuizApp.Entities;
using QuizApp.BLL.Dto.Test;

namespace QuizApp.BLL.Services
{
    public class UrlService : IUrlService
    {
        private readonly IUnitOfWork unitOfWork;

        private readonly ITestRepository testRepository;

        private readonly IUrlRepository urlRepository;

        private readonly ITestResultRepository testResultRepository;

        private readonly IMapper mapper;


        public UrlService(IUnitOfWork unitOfWork, IMapper mapper)
        {
            this.unitOfWork = unitOfWork ?? throw new ArgumentNullException(nameof(unitOfWork));
            this.testRepository = unitOfWork.GetRepository<Test, ITestRepository>() ?? throw new NullReferenceException(nameof(testRepository));
            this.urlRepository = unitOfWork.GetRepository<Url, IUrlRepository>() ?? throw new NullReferenceException(nameof(urlRepository));
            this.testResultRepository = unitOfWork.GetRepository<TestResult, ITestResultRepository>() ?? throw new NullReferenceException(nameof(testResultRepository));
            this.mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));
        }


        public UrlsApiDto GetUrls(int page, int amountUrlsPerPage)
        {
            var urls = urlRepository.GetPageWithAmount(page: page, amountPerPage: amountUrlsPerPage);

            return new UrlsApiDto
            {
                Urls = mapper.Map<List<UrlDto>>(urls),
                TotalCount = urlRepository.Count()
            };
        }

        public async Task<UrlDetailDto> GetUrlById(int urlId)
        {
            Url url = await urlRepository.GetByIdAsync(id: urlId, includeProperties: prop => prop
                .Include(u => u.TestResults)
                .Include(u => u.Test));

            return mapper.Map<UrlDetailDto>(url);
        }

        public async Task<CreatedUrlDto> CreateUrl(NewUrlDto newUrlDto)
        {
            Url url = mapper.Map<Url>(newUrlDto);

            urlRepository.Insert(url);
            await unitOfWork.SaveAsync();

            return mapper.Map<CreatedUrlDto>(url);
        }

        public async Task<UpdatedUrlDto> UpdateUrl(UpdateUrlDto updateUrlDto)
        {
            Url url = await urlRepository.GetByIdAsync(updateUrlDto.Id);
            if (url == null)
            {
                return null;
            }

            url = mapper.Map<Url>(updateUrlDto);

            urlRepository.Update(url);
            await unitOfWork.SaveAsync();

            return mapper.Map<UpdatedUrlDto>(url);
        }

        public TestPreviewDto GetTestByUrlId(int urlId)
        {
            Test testByUrlId = testRepository.Get(filter: test => test.Urls.Any(url => url.Id == urlId)).FirstOrDefault();

            return mapper.Map<TestPreviewDto>(testByUrlId);
        }

        public TestResultsApiDto GetTestResultsByUrlId(int urlId, int page, int amountResultsPerPage)
        {
            var testResults = testResultRepository.GetPageWithAmount(filter: r => r.UrlId == urlId, page: page, amountPerPage: amountResultsPerPage);

            return new TestResultsApiDto
            {
                TestResults = mapper.Map<List<TestResultDto>>(testResults),
                TotalCount = testResultRepository.Count(predicate: r => r.UrlId == urlId)
            };
        }
    }
}

File: frontend/src/app/services/url.service.ts
import { Injectable } from '@angular/core';
import { HttpInternalService } from 'src/app/services/http-internal.service';
import { NewUrlDto } from '../models/url/new-url-dto';
import { CreatedUrlDto } from '../models/url/created-url-dto';
import { UpdateUrlDto } from '../models/url/update-url-dto';
import { UpdatedUrlDto } from '../models/url/updated-url-dto';
import { UrlDetailDto } from '../models/url/url-detail-dto';
import { TestPreviewDto } from '../models/test/test-preview-dto';
import { UrlsApiDto } from '../models/url/urls-api-dto';
import { TestResultsApiDto } from '../models/test-result/test-results-api-dto';

@Injectable({
    providedIn: 'root'
})
export class UrlService {
    private routePrefix = '/api/urls';

    constructor(private httpService: HttpInternalService) { }

    public getUrls(page: number, amountUrlsPerPage: number) {
        return this.httpService.getRequest<UrlsApiDto>(`${this.routePrefix}`, {
            page,
            amountUrlsPerPage
        });
    }

    public getUrlById(id: number) {
        return this.httpService.getRequest<UrlDetailDto>(`${this.routePrefix}/${id}`);
    }

    public createUrl(url: NewUrlDto) {
        return this.httpService.postRequest<CreatedUrlDto>(`${this.routePrefix}`, url);
    }

    public updateUrl(url: UpdateUrlDto) {
        return this.httpService.putRequest<UpdatedUrlDto>(`${this.routePrefix}`, url);
    }

    public getTestByUrlId(urlId: number) {
        return this.httpService.getRequest<TestPreviewDto>(`${this.routePrefix}/${urlId}/test`);
    }

    public getTestResultsByUrlId(id: number, page: number, amountResultsPerPage: number) {
        return this.httpService.getRequest<TestResultsApiDto>(`${this.routePrefix}/${id}/results`, {
            page,
            amountResultsPerPage
        });
    }
}
